File: leetcode/ruby/next_permutation.rb
def nextPermutation(nums)
  i = nums.length - 2
  while i >= 0 && nums[i] >= nums[i + 1]
    i -= 1
  end
  if i >= 0
    j = nums.length - 1
    while j >= 0 && nums[j] <= nums[i]
      j -= 1
    end
    nums[i], nums[j] = nums[j], nums[i]
  end
  nums[i + 1..-1].reverse!
end

def reverse(nums, start, end)
  while start < end
    nums[start], nums[end] = nums[end], nums[start]
    start += 1
    end -= 1
  end
end

# test
nums = [1, 2, 3]
nextPermutation(nums)
puts nums.join(',')

File: leetcode/rust/next_permutation.rs
fn nextPermutation(nums: &mut Vec<i32>) {
    let mut i = nums.len() - 1;
    while i > 0 {
        if nums[i - 1] < nums[i] {
            break;
        }
        i -= 1;
    }
    if i == 0 {
        nums.reverse();
        return;
    }
    let mut j = nums.len() - 1;
    while j >= 0 {
        if nums[i - 1] < nums[j] {
            break;
        }
        j -= 1;
    }
    nums.swap(i - 1, j);
    nums[i..].reverse();
}


#[test]
fn test_next_permutation() {
    let mut nums = vec![1, 2, 3];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![1, 3, 2]);
    let mut nums = vec![3, 2, 1];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![1, 2, 3]);
    let mut nums = vec![1, 1, 5];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![1, 5, 1]);
    let mut nums = vec![1, 3, 2];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![2, 1, 3]);
    let mut nums = vec![2, 3, 1];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![3, 1, 2]);
    let mut nums = vec![1, 1, 5];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![1, 5, 1]);
    let mut nums = vec![1, 2, 3, 4];
    nextPermutation(&mut nums);
    assert_eq!(nums, vec![1, 2, 4, 3]);
}